### ISSUE TITLE
IOS XR: XMLCLIError on large XML output

### ISSUE THREAD

**Issue by rlanyi** [napalm-automation/napalm#1342]
### Description of Issue/Question

Trying to call `get_bgp_neighbors` on an ASR9K router running IOS XR 6.1.2 that has a large number (>1500) of BGP sessions always fails with the exception `XMLCLIError`. Another ASR9K with 6.0.2 and _only_ 1100 sessions runs fine.

Please note that this is happening while I'm testing https://github.com/napalm-automation/napalm/pull/1332, without that I'm always running into the problem mentioned in https://github.com/napalm-automation/napalm/issues/1312.

### Did you follow the steps from https://github.com/napalm-automation/napalm#faq

- [x] Yes
- [ ] No

### napalm version

```
napalm @ git+https://github.com/napalm-automation/napalm.git@8066fae30b9311a279b6feace94775971ac43cc7
```

### Network operating system version

```
Cisco IOS XR Software, Version 6.1.2[Default]
```

### Steps to Reproduce the Issue

I'm running the command below in debug mode. The XML response containes a partial XML from somewhere in the middle of the original XML. (The original XML can be generated manually on the router and its size is around 26M for the 1500 BGP sessions. The output on the second router is 18M and the same napalm command runs just fine.)

The response XML from the router is different every time but it's always between 2.5M-3M in size. I've tried it with/without `fast_cli=True` optional arg. I've also played by changing `timeout` and `delay_factor` values directly in the source code (in `pyIOSXR/iosxr.py`) without success.

<details><summary>Napalm results (Click for details)</summary>
<p>

```
root@c0be17fcbac4:/opt/peering-manager# napalm --debug --user root --password xxxx --vendor iosxr --optional_args "ssh_config_file='/root/.ssh/config',fast_cli=True" router1 call get_bgp_neighbors
2020-12-20 21:48:42,200 - napalm - DEBUG - Starting napalm's debugging tool
2020-12-20 21:48:42,200 - napalm - DEBUG - Gathering napalm packages
2020-12-20 21:48:42,200 - napalm - DEBUG - napalm==3.2.0
2020-12-20 21:48:42,200 - napalm - DEBUG - get_network_driver - Calling with args: ('iosxr',), {}
2020-12-20 21:48:42,201 - napalm - DEBUG - get_network_driver - Successful
2020-12-20 21:48:42,201 - napalm - DEBUG - __init__ - Calling with args: (<class 'napalm.iosxr.iosxr.IOSXRDriver'>, 'router1', 'root'), {'password': '*******', 'timeout': 60, 'optional_args': {'ssh_config_file': '/root/.ssh/config', 'fast_cli': True}}
2020-12-20 21:48:42,201 - napalm - DEBUG - __init__ - Successful
2020-12-20 21:48:42,201 - napalm - DEBUG - pre_connection_tests - Calling with args: (<napalm.iosxr.iosxr.IOSXRDriver object at 0x7f4589d946a0>,), {}
2020-12-20 21:48:42,202 - napalm - DEBUG - open - Calling with args: (<napalm.iosxr.iosxr.IOSXRDriver object at 0x7f4589d946a0>,), {}
2020-12-20 21:48:44,928 - napalm - DEBUG - open - Successful
2020-12-20 21:48:44,928 - napalm - DEBUG - connection_tests - Calling with args: (<napalm.iosxr.iosxr.IOSXRDriver object at 0x7f4589d946a0>,), {}
2020-12-20 21:48:44,929 - napalm - DEBUG - get_facts - Calling with args: (<napalm.iosxr.iosxr.IOSXRDriver object at 0x7f4589d946a0>,), {}
2020-12-20 21:48:54,954 - napalm.pyIOSXR.iosxr - DEBUG - <Element Response at 0x7f4589d52840>
2020-12-20 21:48:57,152 - napalm.pyIOSXR.iosxr - DEBUG - <Element Response at 0x7f4589df9600>
2020-12-20 21:48:57,180 - napalm - DEBUG - Gathered facts:
{
    "vendor": "Cisco",
    "os_version": "6.1.2",
    "hostname": "router1",
    "uptime": 24058255,
    "serial_number": "FOX24000",
    "fqdn": "router1",
    "model": "ASR-9006-AC-V2",
    "interface_list": [
        "BVI1",
        "Bundle-Ether1",
        "GigabitEthernet0/1/0/0",
(...)
    ]
}
{
    "vendor": "Cisco",
    "os_version": "6.1.2",
    "hostname": "router1",
    "uptime": 24058255,
    "serial_number": "FOX24000",
    "fqdn": "router1",
    "model": "ASR-9006-AC-V2",
    "interface_list": [
        "BVI1",
        "Bundle-Ether1",
        "GigabitEthernet0/1/0/0",
(...)
    ]
}
2020-12-20 21:48:57,180 - napalm - DEBUG - get_facts - Successful
2020-12-20 21:48:57,180 - napalm - DEBUG - method - Calling with args: (<napalm.iosxr.iosxr.IOSXRDriver object at 0x7f4589d946a0>, 'get_bgp_neighbors'), {}
2020-12-20 21:48:57,180 - napalm - DEBUG - get_bgp_neighbors - Attempting to resolve method
2020-12-20 21:48:57,180 - napalm - DEBUG - get_bgp_neighbors - Attempting to call method with kwargs: {}
2020-12-20 21:48:57,333 - napalm.pyIOSXR.iosxr - DEBUG - <Element Response at 0x7f4589df9380>
2020-12-20 21:49:11,304 - napalm - ERROR - method - Failed: estartTime><LocalRestartTime>120</LocalRestartTime><StalePathTimeout>360</StalePathTimeout>... (partial XML continues for 2.5M-3M) ...<IsLocalAddressConfigured>false</IsLocalAddressConfigured><Conn

================= Traceback =================

Traceback (most recent call last):
  File "/usr/local/bin/napalm", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.9/site-packages/napalm/base/clitools/cl_napalm.py", line 308, in main
    run_tests(args)
  File "/usr/local/lib/python3.9/site-packages/napalm/base/clitools/cl_napalm.py", line 291, in run_tests
    call_getter(device, args.method, **method_kwargs)
  File "/usr/local/lib/python3.9/site-packages/napalm/base/clitools/cl_napalm.py", line 27, in wrapper
    r = func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/napalm/base/clitools/cl_napalm.py", line 255, in call_getter
    r = func(**kwargs)
  File "/usr/local/lib/python3.9/site-packages/napalm/iosxr/iosxr.py", line 412, in get_bgp_neighbors
    result_tree = ETREE.fromstring(self.device.make_rpc_call(rpc_command))
  File "/usr/local/lib/python3.9/site-packages/napalm/pyIOSXR/iosxr.py", line 172, in make_rpc_call
    result = self._execute_rpc(rpc_command)
  File "/usr/local/lib/python3.9/site-packages/napalm/pyIOSXR/iosxr.py", line 417, in _execute_rpc
    response = self._send_command(xml_rpc_command, delay_factor=delay_factor)
  File "/usr/local/lib/python3.9/site-packages/napalm/pyIOSXR/iosxr.py", line 399, in _send_command
    return self._send_command(
  File "/usr/local/lib/python3.9/site-packages/napalm/pyIOSXR/iosxr.py", line 403, in _send_command
    raise XMLCLIError(output.strip(), self)
napalm.pyIOSXR.exceptions.XMLCLIError: (same partial XML as above)
```
</p>
</details>

#### Router logs

It might be important that while getting the error above the following message appears in the router logs. This is always shown.

```
Dec 20 21:33:06.751 router1 2863031 SECURITY SSHD ERR_GENERAL SSHD_[65918] - error transmitting exec output to network 
```

Sometimes (but not always, seems to be random) it is preceeded with the following:

```
Dec 20 21:34:21.513 router1 2863030 SECURITY SSHD INFO_GENERAL SSHD_[65918] - Client closes socket connection 
```

### Additional info

This is the XML TTY part of my IOS XR config. Changing memory limit or streaming size seems to have no impact on the result, it always fails the same way.

```
xml agent tty
 throttle memory 100
 iteration off
 streaming on size 32
```

**Comment by ktbyers** [comment]
@rlanyi Can you post the full exception stack trace?

**Comment by rlanyi** [comment]
@ktbyers I've posted the stack trace above under "Napalm results", please click it to see the actual results. If you need more details please let me know which CLI parameters should I use to get those.

**Comment by ktbyers** [comment]
I would probably try setting `fast_cli=False` as an argument in optional_args. 

I assume you are using Netmiko 3.3.2? If so, you need to set fast_cli=False in order for delay_factor to work. Once you set fast_cli=False and then I would try setting delay_factor=1 (which might require some modifications to the napalm-pyiosxr code...looks like _READ_DELAY would need set to 1)

If you do this, then you should have 100 seconds for the command to complete.

I am going to try to clean the Netmiko part of this up in Netmiko 4, but some of the pyiosxr items also probably need cleaned up as well.

Let me know if that works.


Regards,

Kirk

**Comment by rlanyi** [comment]
Yes, I'm using Netmiko 3.3.2.

I'm quite sure that I've found the problem: when I played with the `fast_cli` argument I'd always set it to `True` or didn't use it at all. Netmiko's documentation states that the default value for this argument is `False`, but now that you've recommended to set it explicitly to `False` I've tried this and out of 10 tries I got successful result every time.

I've checked Netmiko's source and it seems that `cisco/cisco_xr.py` overrides the default behavior and sets `fast_cli=True` if it is not explicitly set to false from the CLI. The same happens with IOS and NXOS, too.

So there's no real bug here, but I think it'd worth to point this out somewhere in the documentation.
